include/openssl/err.h

```diff
@@ -214,33 +214,33 @@ struct err_state_st {
 
 # define ERR_SYSTEM_ERROR(errcode)      (((errcode) & ERR_SYSTEM_FLAG) != 0)
 
-static ossl_inline int ERR_GET_LIB(unsigned long errcode)
+static ossl_unused ossl_inline int ERR_GET_LIB(unsigned long errcode)
 {
     if (ERR_SYSTEM_ERROR(errcode))
         return ERR_LIB_SYS;
     return (errcode >> ERR_LIB_OFFSET) & ERR_LIB_MASK;
 }
 
-static ossl_inline int ERR_GET_FUNC(unsigned long errcode ossl_unused)
+static ossl_unused ossl_inline int ERR_GET_FUNC(unsigned long errcode ossl_unused)
 {
     return 0;
 }
 
-static ossl_inline int ERR_GET_RFLAGS(unsigned long errcode)
+static ossl_unused ossl_inline int ERR_GET_RFLAGS(unsigned long errcode)
 {
     if (ERR_SYSTEM_ERROR(errcode))
         return 0;
     return errcode & (ERR_RFLAGS_MASK << ERR_RFLAGS_OFFSET);
 }
 
-static ossl_inline int ERR_GET_REASON(unsigned long errcode)
+static ossl_unused ossl_inline int ERR_GET_REASON(unsigned long errcode)
```

Comment:
I must just not be understanding what this actually does..
ssl_rsa.c has this code in it..

&& ERR_GET_REASON(err) == PEM_R_NO_START_LINE)

So is this just saying because it is includes everywhere and its inline it isnt being used in every file that includes it?

Reply:
Right, the potential issue isn't that the defined function is GLOBALLY unused, it's that it might not be used within some particular translation unit. I imagine it's more likely to be an issue for client code pulling in the public headers than within OpenSSL itself.

I believe that typically compilers will ignore inline functions in the context of -Wno-unused-functions. However, there could be cases where "ossl_inline" doesn't expand. AFAIK -Wno-unused-functions would still be supported by the compiler even in cases where inlining isn't.

The linked [issue 12792](https://github.com/openssl/openssl/issues/12792) shows a case where -Wno-unused-functions ends up getting triggered by the functions in this particular header in client (bind9) code. I neglected to include the full command line on the compiler there, so perhaps the full set of flags is a bit "unusual" in some regard... but everything else seemed to build fine and the compiler itself didn't complain about contradictory inputs, so I take that to mean it was fundamentally valid.

My assumption is that the general philosophy in the project is to allow the OpenSSL public headers to "play nice" with as much as possible, provided it doesn't significantly add to the project's own technical burden. In this case, we already had a #define for trying to suppress unused symbol warnings from the compiler, so I'd argue that adding the annotation isn't really adding to the project's technical burden in any meaningful way.